Compilation failed on Alpine distribution

Hello, first of all, thanks for this awesome library!

I tried to compile your extension on the docker image php:7.0-fpm-alpine (which is based on the last alpine distribution ) 
And after installing the zlib-dev package as requested, this extension won't compile.

Dockerfile:
```
FROM php:7.0-fpm-alpine

RUN apk update \
    && apk add --no-cache zlib-dev

## Install spx extension

# git clone of the project
COPY ./spx /usr/src/php/ext/spx 
COPY ./spx.ini /usr/local/etc/php/conf.d/

# run the classic extension installation (same one as indicated in the documentation)
RUN docker-php-ext-install pdo pdo_mysql spx

EXPOSE 9000

CMD ["php-fpm"]
```

As you see, this is pretty straightforward. pdo and pdo_mysql are correctly compiled and installed.

Here is the build report:

```
+ cd spx
+ phpize
Configuring for:
PHP Api Version:         20151012
Zend Module Api No:      20151012
Zend Extension Api No:   320151012
+ ./configure --build=x86_64-linux-musl
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
expr: syntax error
sh: 0: unknown operand
expr: syntax error
sh: 0: unknown operand
expr: syntax error
sh: 0: unknown operand
expr: syntax error
sh: 0: unknown operand
expr: syntax error
sh: 0: unknown operand
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-pc-linux-musl
checking host system type... x86_64-pc-linux-musl
checking target system type... x86_64-pc-linux-musl
checking for PHP prefix... /usr/local
checking for PHP includes... -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib
checking for PHP extension directory... /usr/local/lib/php/extensions/no-debug-non-zts-20151012
checking for PHP installed headers prefix... /usr/local/include/php
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... re2c
checking for re2c version... 0.14.3 (ok)
checking for gawk... no
checking for nawk... no
checking for awk... awk
checking if awk is broken... no
checking whether to enable SPX extension... yes, shared
checking for zlib header... yes
checking for ld used by cc... /usr/x86_64-alpine-linux-musl/bin/ld
checking if the linker (/usr/x86_64-alpine-linux-musl/bin/ld) is GNU ld... yes
checking for /usr/x86_64-alpine-linux-musl/bin/ld option to reload object files... -r
checking for BSD-compatible nm... /usr/bin/nm -B
checking whether ln -s works... yes
checking how to recognize dependent libraries... pass_all
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking dlfcn.h usability... yes
checking dlfcn.h presence... yes
checking for dlfcn.h... yes
checking the maximum length of command line arguments... 98304
checking command to parse /usr/bin/nm -B output from cc object... ok
checking for objdir... .libs
checking for ar... ar
checking for ranlib... ranlib
checking for strip... strip
checking if cc supports -fno-rtti -fno-exceptions... no
checking for cc option to produce PIC... -fPIC
checking if cc PIC flag -fPIC works... yes
checking if cc static flag -static works... yes
checking if cc supports -c -o file.o... yes
checking whether the cc linker (/usr/x86_64-alpine-linux-musl/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no

creating libtool
appending configuration tag "CXX" to libtool
configure: creating ./config.status
config.status: creating config.h
/bin/sh /usr/src/php/ext/spx/libtool --mode=compile cc  -I. -I/usr/src/php/ext/spx -DPHP_ATOM_INC -I/usr/src/php/ext/spx/include -I/usr/src/php/ext/spx/main -I/usr/src/php/ext/spx -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib  -fstack-protector-strong -fpic -fpie -O2 -DHAVE_CONFIG_H  -fstack-protector-strong -fpic -fpie -O2 -Werror -Wfatal-errors -Wall   -c /usr/src/php/ext/spx/src/php_spx.c -o src/php_spx.lo
mkdir src/.libs
 cc -I. -I/usr/src/php/ext/spx -DPHP_ATOM_INC -I/usr/src/php/ext/spx/include -I/usr/src/php/ext/spx/main -I/usr/src/php/ext/spx -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -fstack-protector-strong -fpic -fpie -O2 -DHAVE_CONFIG_H -fstack-protector-strong -fpic -fpie -O2 -Werror -Wfatal-errors -Wall -c /usr/src/php/ext/spx/src/php_spx.c  -fPIC -DPIC -o src/.libs/php_spx.o
In file included from /usr/src/php/ext/spx/src/php_spx.c:1:0:
/usr/src/php/ext/spx/src/php_spx.c: In function 'init':
/usr/src/php/ext/spx/src/php_spx.c:294:23: error: expected identifier before '(' token
     context.fd_backup.stdout = -1;
                       ^
compilation terminated due to -Wfatal-errors.
Makefile:193: recipe for target 'src/php_spx.lo' failed
make: *** [src/php_spx.lo] Error 1
ERROR: Service 'php' failed to build: The command '/bin/sh -c docker-php-ext-install pdo pdo_mysql spx' returned a non-zero code: 2
```

--- a/src/php_spx.c
+++ b/src/php_spx.c
@@ -35,8 +35,8 @@ static SPX_THREAD_TLS struct {
     int cli_sapi;
 
     struct {
-        int stdout;
-        int stderr;
+        int stdout_fd;
+        int stderr_fd;
     } fd_backup;
 
     spx_config_t config;
@@ -291,8 +291,8 @@ static void init(void)
 
     context.cli_sapi = 0 == strcmp(sapi_module.name, "cli");
 
-    context.fd_backup.stdout = -1;
-    context.fd_backup.stderr = -1;
+    context.fd_backup.stdout_fd = -1;
+    context.fd_backup.stderr_fd = -1;
     
     context.profiler = NULL;
     context.output_file[0] = 0;
@@ -334,9 +334,9 @@ static void init(void)
     ) {
         fp_live = context.config.fp_live && isatty(STDOUT_FILENO);
         if (fp_live) {
-            context.fd_backup.stdout = spx_stdio_disable(STDOUT_FILENO);
-            context.fd_backup.stderr = spx_stdio_disable(STDERR_FILENO);
-            output = spx_output_stream_dopen(context.fd_backup.stdout, 0);
+            context.fd_backup.stdout_fd = spx_stdio_disable(STDOUT_FILENO);
+            context.fd_backup.stderr_fd = spx_stdio_disable(STDERR_FILENO);
+            output = spx_output_stream_dopen(context.fd_backup.stdout_fd, 0);
         } else {
             output = spx_output_stream_dopen(STDOUT_FILENO, 0);
         }
@@ -439,12 +439,12 @@ static void finish(void)
     spx_profiler_destroy(context.profiler);
     context.profiler = NULL;
 
-    if (context.fd_backup.stdout != -1) {
-        spx_stdio_restore(STDOUT_FILENO, context.fd_backup.stdout);
+    if (context.fd_backup.stdout_fd != -1) {
+        spx_stdio_restore(STDOUT_FILENO, context.fd_backup.stdout_fd);
     }
 
-    if (context.fd_backup.stderr != -1) {
-        spx_stdio_restore(STDERR_FILENO, context.fd_backup.stderr);
+    if (context.fd_backup.stderr_fd != -1) {
+        spx_stdio_restore(STDERR_FILENO, context.fd_backup.stderr_fd);
     }
 
     spx_resource_stats_shutdown();

--- a/src/spx_reporter_fp.c
+++ b/src/spx_reporter_fp.c
@@ -1,9 +1,9 @@
-#define _GNU_SOURCE /* qsort_r */
 #include <stdio.h>
 #include <stdlib.h>
 #include <time.h>
 
 #include "spx_reporter_fp.h"
+#include "spx_thread.h"
 
 typedef struct {
     spx_profiler_reporter_t base;
@@ -19,10 +19,13 @@ typedef struct {
     const spx_profiler_func_table_entry_t ** top_entries;
 } fp_reporter_t;
 
+static const SPX_THREAD_TLS fp_reporter_t * entry_cmp_reporter;
+
 static spx_profiler_reporter_cost_t fp_notify(spx_profiler_reporter_t * reporter, const spx_profiler_event_t * event);
 static void fp_destroy(spx_profiler_reporter_t * reporter);
 
-static int entry_cmp(const void * a, const void * b, void * data);
+static int entry_cmp(const void * a, const void * b);
+static int entry_cmp_r(const void * a, const void * b, const fp_reporter_t * reporter);
 static size_t print_report(fp_reporter_t * reporter, const spx_profiler_event_t * event);
 
 spx_profiler_reporter_t * spx_reporter_fp_create(
@@ -107,10 +110,25 @@ static void fp_destroy(spx_profiler_reporter_t * reporter)
     free(fp_reporter->top_entries);
 }
 
-static int entry_cmp(const void * a, const void * b, void * data)
+static int entry_cmp(const void * a, const void * b)
 {
-    const fp_reporter_t * reporter = data;
+    /*
+     *  The use of entry_cmp_reporter TLS variable is required as a workaround for the
+     *  absence of qsort_r in some platform.
+     *
+     *  N.B.: This workaround is not reentrant since reentrancy is not required, it
+     *        only have to be thread-safe.
+     */
+    if (!entry_cmp_reporter) {
+        fprintf(stderr, "entry_cmp_reporter is not set\n");
+        exit(1);
+    }
 
+    return entry_cmp_r(a, b, entry_cmp_reporter);
+}
+
+static int entry_cmp_r(const void * a, const void * b, const fp_reporter_t * reporter)
+{
     const spx_profiler_func_table_entry_t * entry_a = (*((const spx_profiler_func_table_entry_t **) a));
     const spx_profiler_func_table_entry_t * entry_b = (*((const spx_profiler_func_table_entry_t **) b));
 
@@ -156,7 +174,7 @@ static size_t print_report(fp_reporter_t * reporter, const spx_profiler_event_t 
         const spx_profiler_func_table_entry_t * current = &event->func_table.entries[i];
         size_t j;
         for (j = 0; j < limit; j++) {
-            if (entry_cmp(&reporter->top_entries[j], &current, reporter) > 0) {
+            if (entry_cmp_r(&reporter->top_entries[j], &current, reporter) > 0) {
                 reporter->top_entries[j] = current;
 
                 break;
@@ -164,12 +182,18 @@ static size_t print_report(fp_reporter_t * reporter, const spx_profiler_event_t 
         }
     }
 
-    qsort_r(
+    /*
+     *  This side effect is required as a workaround for the absence of qsort_r
+     *  in some platform.
+     *  See related comment in entry_cmp implementation.
+     */
+    entry_cmp_reporter = reporter;
+
+    qsort(
         reporter->top_entries,
         limit,
         sizeof(*reporter->top_entries),
-        entry_cmp,
-        reporter
+        entry_cmp
     );
 
     spx_output_stream_print(reporter->base.output, "\n*** SPX Report ***\n\nGlobal stats:\n\n");

